std/assembly/array.ts

```diff
@@ -493,6 +493,68 @@ export class Array<T> {
     return <string>unreachable();
   }
 
+  flat(): T {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
+    }
+    // Get the length and data start values
+    var length = this.length_;
+    var selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    var size = 0;
+    for (let i = 0; i < length; i++) {
+      let child = load<usize>(selfDataStart + (i << alignof<T>()));
+      size += child == 0 ? 0 : load<i32>(child, offsetof<T>("length_"));
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    var byteLength = <usize>size << usize(alignof<valueof<T>>());
+    var dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value and initialize it
+    var result = __alloc(offsetof<T>(), idof<T>());
+    store<i32>(result, size, offsetof<T>("length_"));
+
+    // byteLength, dataStart, and buffer are all readonly
+    store<i32>(changetype<usize>(result), byteLength, offsetof<T>("byteLength"));
+    store<usize>(changetype<usize>(result), dataStart, offsetof<T>("dataStart"));
+    store<usize>(changetype<usize>(result), __retain(dataStart), offsetof<T>("buffer"));
+
+    // set the elements
+    var resultIndex = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child = load<usize>(selfDataStart + (i << alignof<T>()));
+
+      // ignore null arrays
+      if (child == 0) continue;
+
+      // move to the next array index
+      resultIndex++;
+
+      // copy the underlying buffer data to the result buffer
+      let childDataLength = load<i32>(child, offsetof<T>("byteLength"));
+      memory.copy(
+        dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())),
```

Comment:
```suggestion
        dataStart + (<usize>resultIndex << alignof<valueof<T>>()),
```

Reply:
This change was to prevent an IDE error.

Reply:
Please see discussion above.